### PR TITLE
[RQ-630] refactor: persona survey 

### PR DESCRIPTION
**Pull request by RuntimeTerror10**
Closes issue: 

## 📜 Summary of changes:
**Context:**
Till now we show the current survey page based on the page number ( or the index value) and there are various checks on top level to show a particular persona page based on the page number (if page === 2 show recommendation screen). but as new questions are added/removed all these checks also needs to be updated and can break the flow of users who are in the middle of survey.

 
* Track and render survey pages based page id and not page number
* Backwards compatible for old persona survey flow
* code cleanup and refactor

## ✅ Checklist:

- [ ] Make sure linting and unit tests pass.
- [ ] No install/build warnings introduced.
- [ ] Verified UI in browser.
- [ ] For UI changes, added/updated analytics events (if applicable).
- [ ] For changes in extension's code, manually tested in Chrome and Firefox.
- [ ] For changes in extension's code, both MV2 and MV3 are covered.
- [ ] Added/updated unit tests for this change.
- [ ] Raised pull request to update corresponding documentation (if already exists).

### CODE REVIEW

File: app/src/components/misc/PersonaSurvey/ModalFooter.tsx [hunk 1]
```diff
@@ -1,88 +1,84 @@
...
-    if (userPersona[OptionsConfig[currentQuestionnaire as QuestionnaireType]?.key]?.length) return false;
+  const disableContinue = useMemo(() => {
+    if (currentPage === SurveyPage.GETTING_STARTED) return false;
+    if (userPersona[currentQuestionnaire as QuestionnaireType]?.length) return false;
     return true;
-  };
+  }, [currentPage, currentQuestionnaire, userPersona]);
 
   const handleMoveToNextPage = () => {
+    const index = Object.keys(SurveyConfig).indexOf(currentPage);
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/ModalFooter.tsx:39]: how `index` is helping in navigation?

File: app/src/components/misc/PersonaSurvey/ModalFooter.tsx [hunk 2]
```diff
@@ -1,88 +1,84 @@
...
           <Col>
-            {isSurveyModal && currentPage === 0 ? (
+            {isSurveyModal && currentPage === SurveyPage.GETTING_STARTED ? (
               <>
                 <span className="survey-modal-emoji">😁</span> We are excited to see you here
               </>
             ) : (
               <>
-                {currentPage} / {SurveyConfig?.length - 1}
+                {Object.keys(SurveyConfig).indexOf(currentPage)} / {surveyLength - 1}
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/ModalFooter.tsx:74]: can we get the `index` for currentPage only once at top
```suggestion
const currentPageIndex = useMemo(() => Object.keys(SurveyConfig).indexOf(currentPage), [currentPage])
```

File: app/src/components/misc/PersonaSurvey/Option/index.tsx [hunk 3]
```diff
@@ -1,83 +1,30 @@
...
-import { ActiveProps, Option, multipleChoiceOption } from "../types";
 
 interface OptionProps {
   option: Option;
-  questionType: "single" | "multiple";
-  isActive?: (props: ActiveProps) => boolean;
-  action?: (dispatch: any, value: string, clear: boolean, optionType?: string) => void;
-  fieldKey: string;
+  questionnaire: QuestionnaireType;
+  action?: (dispatch: any, value: string, doClear: boolean) => void;
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/Option/index.tsx:10]: can we rename this prop, `action` is too generic

File: app/src/components/misc/PersonaSurvey/config.tsx [hunk 4]
```diff
@@ -1,13 +1,10 @@
...
 import { setUserPersona } from "./actions";
 import { GettingStartedWithSurvey } from "./GettingStartedWithSurvey";
 
 export const OptionsConfig: Record<QuestionnaireType, SurveyOptionsConfig> = {
   [QuestionnaireType.PERSONA]: {
-    key: "persona",
-    questionType: "single",
-    isActive: ({ key, title }) => key === title,
-    action: (dispatch, value, clear) => setUserPersona(dispatch, value, clear, "persona"),
+    action: (dispatch, value, doClear) => setUserPersona(dispatch, value, doClear, QuestionnaireType.PERSONA),
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/config.tsx:7]: can we please rename `value` and `doClear` parameters or add a doc comment about the parameters in type definition

File: app/src/components/misc/PersonaSurvey/index.tsx [hunk 5]
```diff
@@ -24,109 +27,104 @@ interface SurveyProps {
...
-  const shuffledQ1 = useMemo(() => {
-    return shuffleOptions(OptionsConfig[QuestionnaireType.PERSONA].options);
-  }, []);
+  const shuffledQuestionnaire = useMemo(() => {
+    if (currentPage === SurveyPage.RECOMMENDATIONS || currentPage === SurveyPage.GETTING_STARTED) return;
+    const options = OptionsConfig[currentPage]?.options;
+    return shuffleOptions(options);
+  }, [currentPage]);
 
   const SkippableButton = useCallback(() => {
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/index.tsx:41]: please wrap it inside a `useMemo`

File: app/src/components/misc/PersonaSurvey/index.tsx [hunk 6]
```diff
@@ -24,109 +27,104 @@ interface SurveyProps {
...
-      }
+    (questionnaire: QuestionnaireType) => {
+      if (questionnaire) return renderOptions(shuffledQuestionnaire, questionnaire);
     },
-    [shuffledQ1]
+    [shuffledQuestionnaire]
   );
 
-  const renderOptions = (options: Option[], optionSet: QuestionnaireType) => {
+  const renderOptions = (options: Option[], questionnaire: QuestionnaireType) => {
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/index.tsx:77]: use `useMemo`

File: app/src/components/misc/PersonaSurvey/index.tsx [hunk 7]
```diff
@@ -24,109 +27,104 @@ interface SurveyProps {
...
+            option={option}
+            action={OptionsConfig[questionnaire].action}
+            questionnaire={questionnaire}
+          />
+        ))}
+      </div>
     );
   };
 
   const renderPage = useCallback(
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/index.tsx:92]: use `useMemo`

File: app/src/components/misc/PersonaSurvey/types.ts [hunk 8]
```diff
@@ -16,35 +16,19 @@ export interface Option {
...
-  pageId: number;
+  page: number;
+  pageId: SurveyPage;
   title: string;
   subTitle: string;
-  /**
-   * If skip true then don't show the question in survey
-   */
-  skip?: boolean;
+  skip?: boolean; // If skip true then don't show the page in survey
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/types.ts:23]: nit: Add comment as a doc comment, helps in documentation

https://github.com/requestly/requestly/assets/61556757/8c357c19-9b48-43a1-a6af-86cc8204f60b

File: app/src/components/misc/PersonaSurvey/types.ts [hunk 9]
```diff
@@ -58,3 +42,15 @@ export enum PersonaType {
...
   PERSONA = "persona",
 }
+
+export enum SurveyPage {
+  GETTING_STARTED = "getting_started",
+  PERSONA = "persona",
+  RECOMMENDATIONS = "recommendations",
+}
+
+export const OldSurveyPageMap = {
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/types.ts:52]: nit: constant is defined inside the type file

**Review by rohanmathur91 — CHANGES_REQUESTED** [review]

File: app/src/components/misc/PersonaSurvey/ModalFooter.tsx (hunk 1 above)
Comment (RuntimeTerror10) [reply app/src/components/misc/PersonaSurvey/ModalFooter.tsx:39]: index is used here to make decisions on last step of persona survey, wether to move to recommendation screen or not , or the survey is from workspace onboarding then perform its necessary actions on survey completion.

File: app/src/components/misc/PersonaSurvey/index.tsx [hunk 10]
```diff
@@ -24,109 +27,107 @@ interface SurveyProps {
...
-              Sign in
-            </RQButton>
-          </div>
-        );
-      default:
-        return null;
-    }
-  }, [currentPage, dispatch]);
+  const renderSkipButton = useMemo(() => {
+    return () => {
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/index.tsx:42]: no need to return a component, directly return JSX
```jsx
const skipButton = useMemo(() => <>....</>, [dispatch])
```

File: app/src/components/misc/PersonaSurvey/index.tsx [hunk 11]
```diff
@@ -24,109 +27,107 @@ interface SurveyProps {
...
+          Sign in
+        </RQButton>
+      </div>;
+    };
+  }, [dispatch]);
 
-  const renderPageHeader = useCallback(
-    (page: PageConfig) => {
+  const renderOptions = useMemo(
+    () => (options: Option[], questionnaire: QuestionnaireType) => {
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/index.tsx:71]: same as above

File: app/src/components/misc/PersonaSurvey/index.tsx [hunk 12]
```diff
@@ -24,109 +27,107 @@ interface SurveyProps {
...
-      </>
-    );
-  };
+      );
+    },
+    []
+  );
+
+  const renderQuestionnaire = useMemo(
+    () => (questionnaire: QuestionnaireType) => {
```
Comment (rohanmathur91) [inline app/src/components/misc/PersonaSurvey/index.tsx:89]: https://github.com/requestly/requestly/pull/829/files#r1273035923

**Review by rohanmathur91 — APPROVED** [review]

**Pull request merged**